// @storybook/ui/dist/modules/ui/components/left_panel/stories_tree/tree_decorators.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

exports.default = function (parent) {
  return (0, _extends3.default)({}, _reactTreebeard.decorators, {
    Header: createHeaderDecoratorScope(parent),
    Container: ContainerDecorator,
    Toggle: ToggleDecorator
  });
};

var _reactTreebeard = require('react-treebeard');

var _io = require('react-icons/lib/io');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ToggleDecorator(_ref) {
  var style = _ref.style;
  var height = style.height,
      width = style.width,
      arrow = style.arrow;


  return _react2.default.createElement(
    'div',
    { style: style.base },
    _react2.default.createElement(
      'div',
      { style: style.wrapper },
      _react2.default.createElement(_io.IoChevronRight, { height: height, width: width, style: arrow })
    )
  );
}

ToggleDecorator.propTypes = {
  style: _propTypes2.default.shape({
    width: _propTypes2.default.number.isRequired,
    height: _propTypes2.default.number.isRequired,
    arrow: _propTypes2.default.object.isRequired
  }).isRequired
};

function ContainerDecorator(props) {
  var node = props.node;


  if (node.root) {
    return null;
  }

  return _react2.default.createElement(_reactTreebeard.decorators.Container, props);
}

ContainerDecorator.propTypes = {
  node: _propTypes2.default.shape({
    root: _propTypes2.default.bool
  }).isRequired
};

function createHeaderDecoratorScope(parent) {
  var HeaderDecorator = function (_React$Component) {
    (0, _inherits3.default)(HeaderDecorator, _React$Component);

    function HeaderDecorator() {
      var _ref2;

      (0, _classCallCheck3.default)(this, HeaderDecorator);

      for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      var _this = (0, _possibleConstructorReturn3.default)(this, (_ref2 = HeaderDecorator.__proto__ || (0, _getPrototypeOf2.default)(HeaderDecorator)).call.apply(_ref2, [this].concat(args)));

      _this.onKeyDown = _this.onKeyDown.bind(_this);
      return _this;
    }

    (0, _createClass3.default)(HeaderDecorator, [{
      key: 'onKeyDown',
      value: function onKeyDown(event) {
        var onKeyDown = parent.onKeyDown;
        var node = this.props.node;


        onKeyDown(event, node);
      }
    }, {
      key: 'render',
      value: function render() {
        var _props = this.props,
            style = _props.style,
            node = _props.node;


        var newStyleTitle = (0, _extends3.default)({}, style.title);

        if (!node.children || !node.children.length) {
          newStyleTitle.fontSize = '13px';
        }

        return _react2.default.createElement(
          'div',
          { style: style.base, role: 'menuitem', tabIndex: '0', onKeyDown: this.onKeyDown },
          _react2.default.createElement(
            'a',
            { style: newStyleTitle },
            node.name
          )
        );
      }
    }]);
    return HeaderDecorator;
  }(_react2.default.Component);

  HeaderDecorator.propTypes = {
    style: _propTypes2.default.shape({
      title: _propTypes2.default.object.isRequired,
      base: _propTypes2.default.object.isRequired
    }).isRequired,
    node: _propTypes2.default.shape({
      name: _propTypes2.default.string.isRequired
    }).isRequired
  };

  return HeaderDecorator;
}
